### PR TITLE
Support for multiple yaml documents in stdin/file

### LINKED ISSUE #10866: Support multiple YAML documents through stdin
Hi,

as maintainer of https://github.com/jkroepke/helm-secrets, I'm looking into a way to pass decrypted value secured files to helm.

Currently, encrypted value files needs to be decrypted first and the decrypted value file needs to be passed 2 a second helm call.

Helm allow values from stdin (https://github.com/helm/helm/pull/2809), but it's not possible to pass multiple YAML documents through stdin.

### PR DESCRIPTION
**Pull request by jkroepke**
**What this PR does / why we need it**:

Fixes #10866 

**Special notes for your reviewer**:

Tested stdin yaml:

```yaml
---
a:
  b: c
  c: d
  e: f
---
e: value
a:
  b: f
  e:
  
test: >-
  hello
  ---
  world
---
a:
  c: 4
```

Before (main branch):
```
% cat values.yaml | ./bin/helm template --repo https://jkroepke.github.io/helm-charts/ values -f - 
---
# Source: values/templates/values.yaml
a:
  b: c
  c: d
  e: f
```

After (this PR):
```
% cat values.yaml | ./bin/helm template --repo https://jkroepke.github.io/helm-charts/ values -f - 
---
# Source: values/templates/values.yaml
a:
  b: f
  c: 4
  e: null
e: value
test: hello --- world
```

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by joejulian** [comment]
Can I suggest a different approach to Unmarshalling multi-document yaml? What about using this example? https://go.dev/play/p/MZNwxdUzxPo

**Comment by joejulian** [comment]
I've tried the method you're suggesting, in the past, and something caused it to break mid-file. I can't remember what it was though.

**Comment by jkroepke** [comment]
> What about using this example?

It uses a different golang yaml library (`sigs.k8s.io/yaml` vs `goyaml "github.com/go-yaml/yaml"`) and uses the v1?. No idea, if this is wanted

Kubernetes uses the same separater ( https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go#L199). I got this from here: https://github.com/kubernetes-sigs/yaml/issues/37#issuecomment-933922133

**Comment by jkroepke** [comment]
I provide different versions based on your example:

* https://github.com/jkroepke/helm/blob/857fd1286f3671a7bf07f5bee9c497ecf2b26ac9/pkg/cli/values/options.go#L54-L70 
  (from https://github.com/kubernetes-sigs/yaml/issues/37#issuecomment-1098693780)
* https://github.com/jkroepke/helm/blob/36c552bb5b0d94d3884d1fe1c4cc173784a00b4f/pkg/cli/values/options.go#L55-L77
  (from https://github.com/kubernetes/kubernetes/blob/ddf47ac13c1a9483ea035a79cd7c10005ff21a6d/cmd/kubeadm/app/util/marshal.go#L80-L97)

**Comment by jkroepke** [comment]
Based on https://github.com/kubernetes-sigs/yaml/issues/37#issuecomment-1098693780, I push a new version here.

**Review by joejulian — APPROVED** [review]
This looks great, to me!

I am not a core maintainer, though. This will still need their approval.

**Comment by jkroepke** [comment]
@thomastaylor312 @adamreese What did you think?

**Comment by jkroepke** [comment]
@joejulian How can I help here? Looks stale here.

**Comment by joejulian** [comment]
@jkroepke I've been wondering that, too. Maybe the helm mailing list or, if you can attend, the [helm developer call](https://docs.google.com/document/d/1d-6xJEx0C78csIYSPKJzRPeWaHG_8W1Hjl72OJggwdc/edit)?

**Comment by jkroepke** [comment]
@joejulian thanks. https://lists.cncf.io/g/cncf-helm/message/450

**Comment by jkroepke** [comment]
@mattfarina What did you think?

**Comment by yxxhero** [comment]
@jkroepke I think you shuld write a HIP for this.

**Comment by jkroepke** [comment]
Any news here @yxxhero ?

**Comment by jkroepke** [comment]
Is it possible to do a review here soon? 

Maybe from @adamreese @thomastaylor312 @mattfarina @hickeyma or @scottrigby ?

**Comment by nce** [comment]
Could someone update this issue, why it's not progressing? Is there anything missing? We rely on `helm-secrets` and the usecase this change supplies

**Comment by yxxhero** [comment]
@jkroepke I will see this PR in the weekend.

File: pkg/cli/values/options.go
Comment (yxxhero) [inline pkg/cli/values/options.go:77]: why changes?

File: pkg/cli/values/options.go
Comment (yxxhero) [inline pkg/cli/values/options.go:90]: add more comments

File: pkg/cli/values/options.go
Comment (jkroepke) [reply pkg/cli/values/options.go:77]: Variable 'bytes' collides with imported package name

File: pkg/cli/values/options.go
Comment (yxxhero) [inline pkg/cli/values/options.go:90]: golang --> Go

**Comment by yxxhero** [comment]
@jkroepke I think the code is good.

**Review by yxxhero — APPROVED** [review]
/lgtm

**Comment by yxxhero** [comment]
@hickeyma @marckhouzam @scottrigby  Can you have a review on this. Thanks very much.

**Comment by iverberk** [comment]
@hickeyma @marckhouzam @scottrigby If you have some time, can you please review and (if ok) merge this addition? It unlocks a nice workflow to pass secrets securely to Helm (no intermediate disk storage needed).

**Comment by iverberk** [comment]
Friendly reminder to review this PR. It's fine (but slightly disappointing) ff there is no time or urgency to review this PR but it would be nice to get some kind of an update.

**Comment by yxxhero** [comment]
@iverberk I added this to this 3.10.0

**Comment by iverberk** [comment]
Wonderful @yxxhero! Thanks for the update and putting this on a milestone, looking forward to start using it.

**Review by hickeyma — CHANGES_REQUESTED** [review]
@jkroepke There is feedback in the HIP (https://github.com/helm/community/pull/253#pullrequestreview-1062334242) from @bacongobbler. Do you mind addressing the comments?

**Comment by jkroepke** [comment]
Hi @hickeyma,

I read the feedback and there are currently no plan to more forward on the HIP. This PR add 30 lines of code and remove 9 lines of code excluding code for tests. From my point of view, a HIP feels too over engineered for this kind of change. I'm not going to document the existing values merge logic in detail or discuss about security implications which are not in scope of the helm project.

At least, the HIP feedback does not criticism the current implementation draft. I'm not sure, if there is an alternative solution to implement this feature from user perspective. 

From code perspective, there are 2 approvals from 2 triage maintainers. All questions are resolved on this PR expect the lastest one. Unit tests are included. I would really like to move forward on the PR here.

**Comment by hickeyma** [comment]
@jkroepke If this feature requires a HIP then the HIP needs to be agreed upon before the feature code can be merged as the feature implementation needs to follow the design. 

@bacongobbler reviewed the HIP and provided feedback and it is good practice to address these comments first.

**Comment by jkroepke** [comment]
I saw the comments, and in some points I do not agree with him. Maybe someone else from the community has enough motivation for this process.

**Comment by iverberk** [comment]
I agree with @jkroepke that this process seems very heavy-handed for the proposed change. It's basically achieving parity with something that can already be done on the CLI (providing multiple values files). It's also reusing a lot of the existing code. Of course, it's your prerogative to mandate following the HIP process but I think no one will be motivated enough to follow through, especially on the long feedback cycles. Maybe that's also indicative that this is a 'niche' change that doesn't affect a lot of people. That being said, I think the use-case is very valid and that this should be part of core Helm.

@jkroepke I've decided to go ahead and just merge the values files myself before presenting them to Helm. I've taken the mergeMaps function and applied it to the values files that I have loaded, after that it's just a matter of marshalling to bytes and feeding it via stdin. Maybe you can take a similar approach to get what you want (secure passing of values via stdin). Thanks for your efforts on this so far.

**Comment by jkroepke** [comment]
I'm also still confused, why other features does not require a HIP, while this feature does.

https://github.com/helm/helm/pull/10693

I may not see the logic behind this.

**Comment by joejulian** [comment]
Just to be clear, triage folks are not code maintainers. We triage issues. Our feedback on PRs has exactly as much weight as anybody else in the community.

**Pull request closed without merging**
